File: symfony-app/src/Service/ContentRecommendationService.php
<?php

namespace App\Service;

use App\Entity\User;
use App\Repository\ContenuRepository;

class ContentRecommendationService
{
    public function __construct(private ContenuRepository $contenuRepository)
    {
    }

    /**
     * @return array<int, \App\Entity\Contenu>
     */
    public function getRecommendedFor(User $user, int $limit = 6): array
    {
        $limit = max(1, min(12, $limit));
        $keywords = $this->buildKeywords($user);

        if ($keywords !== []) {
            $qb = $this->contenuRepository->createValidatedQueryBuilder();
            $orX = $qb->expr()->orX();
            foreach ($keywords as $index => $keyword) {
                $param = 'k' . $index;
                $orX->add("c.titre LIKE :$param");
                $orX->add("c.description LIKE :$param");
                $orX->add("c.tags LIKE :$param");
                $orX->add("c.categorie LIKE :$param");
                $qb->setParameter($param, '%' . $keyword . '%');
            }

            $qb->andWhere($orX)
                ->orderBy('c.datePublication', 'DESC')
                ->addOrderBy('c.createdAt', 'DESC')
                ->setMaxResults($limit);

            $items = $qb->getQuery()->getResult();
            if (count($items) > 0) {
                return $items;
            }
        }

        return $this->contenuRepository->findValidatedPage(1, $limit)['items'];
    }

    /**
     * @return string[]
     */
    private function buildKeywords(User $user): array
    {
        $keywords = [];
        $role = $user->getRole();
        $subscriptionType = $user->getSubscriptionType();

        $map = [
            'ROLE_PATIENT' => ['patient', 'bien-etre', 'sante'],
            'ROLE_MEDECIN' => ['medecin', 'medical', 'clinique'],
            'ROLE_COACH' => ['sport', 'coaching', 'entrainement'],
            'ROLE_NUTRITIONNISTE' => ['nutrition', 'alimentaire', 'dietetique'],
        ];

        if (isset($map[$role])) {
            $keywords = array_merge($keywords, $map[$role]);
        }
        if ($subscriptionType) {
            $keywords[] = $subscriptionType;
        }

        if ($user->getMedecin()) {
            $keywords[] = $user->getMedecin()->getSpecialite();
        }
        if ($user->getCoachSportif()) {
            $keywords[] = $user->getCoachSportif()->getSpecialite();
        }
        if ($user->getNutritionniste()) {
            $keywords[] = $user->getNutritionniste()->getSpecialite();
        }

        foreach ($user->getLikes() as $like) {
            $contenu = $like->getContenu();
            if ($contenu->getCategorie()) {
                $keywords[] = $contenu->getCategorie();
            }
            if ($contenu->getTags()) {
                $keywords = array_merge($keywords, array_map('trim', explode(',', $contenu->getTags())));
            }
        }

        foreach ($user->getCommentaires() as $commentaire) {
            $contenu = $commentaire->getContenu();
            if ($contenu->getCategorie()) {
                $keywords[] = $contenu->getCategorie();
            }
            if ($contenu->getTags()) {
                $keywords = array_merge($keywords, array_map('trim', explode(',', $contenu->getTags())));
            }
        }

        return array_values(array_unique(array_filter(array_map('strtolower', $keywords))));
    }
}

File: javafx-app/src/main/java/com/santea/service/ContentRecommendationService.java
package com.santea.service;

import com.santea.model.User;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Recommandation de contenu : mots-clés (rôle, abonnement, spécialités, likes / commentaires)
 * + fallback publications récentes — aligné sur symfony-app ContentRecommendationService::buildKeywords.
 */
public final class ContentRecommendationService {
    private ContentRecommendationService() {
    }

    public static List<String> buildKeywords(User user) {
        if (user == null) {
            return List.of();
        }

        List<String> keywords = new ArrayList<>();
        String role = normalize(user.getRole()).toUpperCase(Locale.ROOT);
        String subscriptionType = normalize(user.getSubscriptionType());

        Map<String, List<String>> map = Map.of(
                "ROLE_PATIENT", List.of("patient", "bien-etre", "sante"),
                "ROLE_MEDECIN", List.of("medecin", "medical", "clinique"),
                "ROLE_COACH", List.of("sport", "coaching", "entrainement"),
                "ROLE_NUTRITIONNISTE", List.of("nutrition", "alimentaire", "dietetique")
        );

        keywords.addAll(map.getOrDefault(role, List.of()));
        if (!subscriptionType.isBlank()) {
            keywords.add(subscriptionType.toLowerCase(Locale.ROOT));
        }

        if (user.getMedecin() != null && user.getMedecin().getSpecialite() != null) {
            keywords.add(user.getMedecin().getSpecialite());
        }
        if (user.getCoachSportif() != null && user.getCoachSportif().getSpecialite() != null) {
            keywords.add(user.getCoachSportif().getSpecialite());
        }
        if (user.getNutritionniste() != null && user.getNutritionniste().getSpecialite() != null) {
            keywords.add(user.getNutritionniste().getSpecialite());
        }

        LinkedHashMap<String, Boolean> unique = new LinkedHashMap<>();
        for (String k : keywords) {
            if (k == null) {
                continue;
            }
            String lower = k.trim().toLowerCase(Locale.ROOT);
            if (!lower.isEmpty()) {
                unique.put(lower, Boolean.TRUE);
            }
        }
        return new ArrayList<>(unique.keySet());
    }

    private static String normalize(String value) {
        return value == null ? "" : value.trim();
    }
}
